cmd/snap-update-ns/utils.go

```diff
@@ -79,7 +138,37 @@ type Secure struct{}
 // things in $SNAP_DATA or in /tmp but we don't want to do so in /etc or in
 // other sensitive places.
 func (sec *Secure) CheckTrespassing(fd int, segments []string, segNum int) error {
-	return nil
+	// If the full path is allowed there's no problem. We don't go and check
+	// the filesystem type because it is just going to spam tests with extra
+	// fstatfs calls and it doesn't change anything.
+	pd := "/" + strings.Join(segments, "/") + "/"
+	for _, prefix := range sec.allowedPaths {
+		logger.Debugf("checking string prefix %q %q", pd, prefix)
+		if strings.HasPrefix(pd, prefix) {
+			return nil
+		}
+	}
+
+	// Otherwise it depends on where we are. The filesystem can be either
+	// tmpfs or squashfs. Those are "safe" because tmpfs is usually made by
+	// us for a writable mimic (this is not a perfect check) and squashfs is
+	// always read only so we will fail on the write operation anyway.
+	// All other filesystems are not allowed and will return an error that
+	// triggers mimic construction.
+	//
+	// Note that some paths are always disallowed but this happens earlier
+	// at snap validation stage. This most notably includes /run, which is
+	// a tmpfs filesystem.
+	var fsData syscall.Statfs_t
+	if err := sysFstatfs(fd, &fsData); err != nil {
+		return fmt.Errorf("cannot fstatfs path segment %q: %s", segments[segNum], err)
+	}
+	if fsData.Type == TmpfsMagic || fsData.Type == SquashfsMagic {
```

Comment:
Above it says this is not a perfect check, and lists /run as an example. So what's the plan here?  What if there are other sensitive filesystems in the host?

Reply:
Ohh, this can be fixed perfectly! 

`sudo mount none -t tmpfs ./experiment/ -o x-snapd.role=mimic`

This is stored and explicitly documented as such in mount(8).

@niemeyer can this be a follow-up, I want to think what else we can do with this knowledge!

Reply:
It can be a follow up given that the current behavior is to allow everything, as I understand it, right?

But there are other points still open in this PR.

Reply:
The current behaviour is to allow writes to any tmpfs, yes. This will probably be a bit more complex as we later found out this is stored in userspace, not in the kernel.